core/client_test.go

```diff
@@ -96,6 +98,139 @@ func TestClientImplementsRequestMaker(t *testing.T) {
 // Perhaps consider golden/testdata integration for generating request/response
 // data.
 
+type fakeRequestMaker2Call struct {
+	Ctx     context.Context
+	Request *katapult.Request
+	V       interface{}
+}
+
+type fakeRequestMaker2Responder func(
+	frm *fakeRequestMaker2,
+	ctx context.Context,
+	req *katapult.Request,
+	v interface{},
+) (ok bool, resp *katapult.Response, err error, retV interface{})
+
+func newFakeRequestMaker2AnyResponder(
+	resp *katapult.Response,
+	err error,
+	v interface{},
+) fakeRequestMaker2Responder {
+	return func(
+		_ *fakeRequestMaker2,
+		_ context.Context,
+		_ *katapult.Request,
+		_ interface{},
+	) (bool, *katapult.Response, error, interface{}) {
+		return true, resp, err, v
+	}
+}
+
+type fakeRequestMaker2 struct {
+	T          *testing.T
+	Calls      []fakeRequestMaker2Call
+	Responders []fakeRequestMaker2Responder
+
+	// Convenience fields populated by the very first call to DoIt. To access
+	// subsequent call methods, use the calls fields.
+	Ctx     context.Context
+	Request *katapult.Request
+	V       interface{}
+}
+
+func newFakeRequestMaker2(
+	t *testing.T,
+	resp *katapult.Response,
+	err error,
+	v interface{},
+) *fakeRequestMaker2 {
+	return &fakeRequestMaker2{
+		T: t,
+		Responders: []fakeRequestMaker2Responder{
+			newFakeRequestMaker2AnyResponder(resp, err, v),
+		},
+	}
+}
+
+func (s *fakeRequestMaker2) DoIt(
+	ctx context.Context,
+	req *katapult.Request,
+	v interface{},
+) (*katapult.Response, error) {
+	copy, err := copystructure.Copy(fakeRequestMaker2Call{
+		Ctx:     ctx,
+		Request: req,
+		V:       v,
+	})
+	require.NoError(s.T, err,
+		"could not deep-copy DoIt arguments to static call list",
+	)
+	call := copy.(fakeRequestMaker2Call)
+
+	if len(s.Calls) == 0 {
+		s.Ctx = call.Ctx
+		s.Request = call.Request
+		s.V = call.V
+	}
+
+	s.Calls = append(s.Calls, call)
+
+	// Iterate through responders in reverse order, allowing latest defined to
+	// take precedence.
+	for i := len(s.Responders) - 1; i >= 0; i-- {
+		f := s.Responders[i]
+		ok, resp, err, val := f(s, ctx, req, v)
+		if !ok {
+			continue
+		}
+
+		rVal := reflect.ValueOf(val)
+		if val != nil && !rVal.IsZero() {
```

Comment:
Comment explaining what this section does (within the if statement) would be nice as a lot of engs aren't familiar with reflect.